Tests, implementation and documentation for from_quaternion

adbugger opened the pull request:
The `__init__` function accepts prepared quaternion matrices of size `[N x 4]`. The `from_quaternion` function accepts `numpy.array` of shape `(4, )` or `(N x 4)`, prepares them and calls `__init__`.

Tests have been run with `scipy.spatial.transform.test()`

adbugger committed:
Implemented from_quaternion. Class method not being documented.

adbugger committed:
Added tests for from_quaternion function

adbugger commented:
Right now, I'm trying to figure out why the `from_quaternion` class method is not being documented.

adbugger committed:
Shifted normalisation to `__init__` function

adbugger commented:
On second thought, I realized that `__init__` should also check for unit norm quaternions. Since we'd be calculating L2 norms anyway, it will be better to shift the normalization code to `__init__` itself, rather than `from_quaternion`.
 Now, `from_quaternion` only converts a possible input of shape `(4,)` to `(1 x 4)`

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,55 @@
 from __future__ import division, print_function, absolute_import
 
+__all__ = ['Rotation']
```
It seem like this `__all__` typically is not necessary. I suggest not to include lines without strong reasons.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,55 @@
 class Rotation(object):
     """Rotation in 3 dimensions.
 
     This class will include initializers from different representations, converters
```
Sure this documentation is preliminarily. While we are at it, I think it should explain:

1. What rotation is mathematically.
2. How do we view a rotation (as physical rotation of the original frame to the final frame).
3. As the consequence of 2 if we have the chain of rotations A -> B -> C, the R_AC = R_AB * R_BC

You don't need to focus on this documentation now, but something to keep in mind.

adbugger replied:
I'll keep that in mind. I was planning to finalize documentation once the implementation was done.

larsoner replied:
This is not PEP8 compliant (>80 char). I'd recommend setting up automated PEP8 checking in your editor.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,55 @@
-    def __init__(self):
-        pass
+    def __init__(self, quaternions):
+        """Initialize class from possible normalized quaternions.
```
We need to stick to some strategy for naming. I think perhaps use singular form (i.e. quaternion) and perhaps shorten it to `quat` for arguments. 

I'm also inclined to use `to_quat`, `as_quat`, but let's leave it for now.

adbugger replied:
`quat` for arguments sounds fine.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,55 @@
+
+        Parameters
+        ----------
+        quaternions : (N, 4) numpy.array
```
OK, typically it is written as `array_like, shape (4,) or (n, 4)`. It means that it might be anything convertable to numpy array.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,55 @@
+                      (x, y, z, w) format.
+
+        """
+        # Each row should have 4 numbers
```
So, first you need to convert using `np.asarray`, then check it's shape and raise `ValueError` if something is wrong with it.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,55 @@
+        assert(quaternions.shape[1] == 4)
+
+        # L2 norm of each row
+        norms = scipy.linalg.norm(quaternions, axis=1)
```
I came to the conclusion that we need to pass a flag `normalize=True`. Meaning that if a user guarantees they are normalized we don't do that again. Makes sense to you?

adbugger replied:
Sure, makes sense.
Just a semantic nitpick. How about a `normalized=True` flag? If the input quaternions are already `normalize`__d__, then we do not need to do  anything further.

nmayorov replied:
Haha, I thought about that too. Whatever feels more natural to you, if you choose `normalized`, then it should be `normalized=False`, i. e. by default we are cautious about that. Right?

adbugger replied:
Yes, of course. If it is `normalized`, then default will be `False`. Sorry I forgot to clarify that.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,55 @@
+
+        # Divide each row by its norm.
+        # In case quat is [4 x 4], ensure norm is broadcasted along each column
+        normalised_quat = quaternions / norms[:, None]
```
I guess we need to check for `norms == 0`, in this case raise a warning and set the quaternion to unity. Makes sense?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,55 @@
+        self._quat = normalised_quat
+
+    @classmethod
+    def from_quaternion(cls, quat):
```
I think it would make sense to simply delegate everything to `__init__`. I suggest write the detailed documentation in here and mention `from_quaternion` in `__init__`. What do you think?

adbugger replied:
Sure, that also works.

nmayorov reviewed (commented):
Some thoughts.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,55 @@
+        """
+
+        # If a single quaternion is given, convert it to a 2D 1 x 4 matrix
+        if quat.shape == (4,):
```
Here are my thoughts about storage. So you decided to store quaternions in stack always. What this leads to in my opinion:

1. You need to store a flag internally like `self._single` and if `self._single == True` then return a single object in `as_...`. To clarify, this is required to separate a possible case when a user passes `(1, 4)` array from when he passes `(4,)` array.
2. This part I'm not 100% sure. I would guess that applying vectorized approach for very few objects (especially 1) is not a great idea in terms of performance. On the other hand this is not measured. So what would be your strategy for this: 
    1. Code only vectorized approach.
    1. Code both approaches right away. And then either 1) iterate over objects and apply a scalar approach or 2) apply vectorized approach, all depending on performance benchmarking for different number of rotation objects (like dcm, quat, etc. )
    1. Code one of the approaches first and came back later to implement another if you have time.

I might guess that you should first experiment with it, say, for `from_dcm` code. See the performance difference depending on the number of rotation objects and then choose one of the approaches.

What do you think?

adbugger replied:
1. Can you tell me the purpose of the `_single` flag? In my mind, we can just check the shape of the `_quat` matrix to see how many quaternions are there. (**Edit**: Thanks for the clarification. So we treat `(1 x 4)` and `(4,)` cases separately even though they represent only one quaternion, got it.)

Regarding your second point: (I'm assuming that by _stack_ you are referring to the `_quat` variable and by _objects_ on the stack you are referring to each row of `_quat`? ).

If I understand your thought process correctly, the concern is that vectorized approaches like `scipy.linalg.norm` and broadcasting tricks will have a significant performance overhead when dealing with very few rotations. For this, I propose a `threshold` number of quaternions in a single `Rotation` object. If the number of rotations is more than this threshold, then we apply vectorized operations, otherwise we iterate over all rows in `_quat` and do our operations. This threshold will have to be chosen according to the benchmarks and various operations in our implementation.

What I'm thinking is that I'll implement the vectorized approach for now. Once this function is implemented, I'll perform some benchmarking tests to see if there are performance impacts, and at what number can we safely use vectorized approaches. Is that ok?

nmayorov replied:
> Can you tell me the purpose of the _single flag? In my mind, we can just check the shape of the _quat matrix to see how many quaternions are there.

I added the explanation later, perhaps you didn't see it. So the problematic case is `(1, 4)` input )

> If I understand your thought process correctly, the concern is that vectorized functions like scipy.linalg.norm will have a significant performance overhead when dealing with very few rotations. For this, I propose a threshold number of quaternions in a single Rotation object. If the number of rotations is more than this threshold, then we apply vectorized operations, otherwise we iterate over all rows in _quat and do our operations. This threshold will have to be chosen according to the benchmarks and various operations in our implementation.

Your general understanding is correct. However I'm speaking about other cases. Well, it isn't strongly related to `_quat`, my bad here. 

For `from_quat` I don't see any problem in vectorized approach. But consider `from_dcm` --- as you may recall there are 4 different cases. If you do the vectorized approach you will write something like `mask_1 = (boolean mask for case 1)`, `mask_2 = ...`, `mask_3 = ...`, `mask_4 = ...`. Then you write something like `quat_[mask_1] = ...`.  You can see that this is quite strange and possibly inefficient approach when we have only few dcms to convert. The same issues will arise in other conversion routines.

From our discussion I suggest to solve this case by case. In difficult cases like `from_dcm` I believe both approaches might be necessary. So I suggest to experiment with it in `from_dcm` to see what is the difference in performance.

Is it clear what I said?

adbugger replied:
Oh yes, that is much clearer, thanks. I was aware of the 4 cases, I just thought this concern was related to `_quat`. I'll keep that in mind when I'm writing the `from_dcm` function.

nmayorov commented:
I figured the problems with the docstring:

1. You write `Parameters` for `__init__` in the docstring for the class, you don't use docstring for `__init__`. Explain that you should use `from_...`class methods and `from_quaternion` in particular. I believe we don't need `Parameters` section for `Rotation` class. 
2. List all methods in `Methods` section. 

See the how it is done [here](https://github.com/scipy/scipy/blob/master/scipy/interpolate/_cubic.py#L352) for example.

adbugger committed:
Added `normalized` flag. Class method still not documented.

adbugger commented:
I made the appropriate changes. However, the class method `from_quaternion` is still not being documented for some reason.

nmayorov commented:
Something is not quite right with the documentation build setup. I don't even get `spatial.transform` page now. See here https://github.com/scipy/scipy/tree/master/doc/source --- it contains files for each page that must be generated. You should add `spatial.transform.rst` similar to `spatial.distance.rst`. Then see what happens.

Other than that you need to modify `setup.py` files. See `signal` and `signal/windows` for example.

nmayorov commented:
I tested this recipe and it worked for me. Make this fix and setup.py modifications and move one.

adbugger committed:
Finally fixed documentation. All methods are now documented.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,74 @@
+        # If a single quaternion is given, convert it to a 2D 1 x 4 matrix but
+        # set self._single to True so that we can return appropriate objects
+        # in the `to_...` methods
+        if quat.shape == (4,):
```
I think the checking logic should be slightly different to make it more robust.

1. Check that `quat.ndim in [1, 2]`.
2. If it is 1, broadcast it to 2D.
3. Then your last check applies.

As it is implementing it won't cover cases (as expected) for shapes like `(5,)` or `(6, 4, 10)`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,74 @@
+
+        # Each row should have 4 numbers
+        if quat.shape[1] != 4:
+            raise ValueError("A quaternion should have 4 numbers in (x,y,z,w) format")
```
Just as a sort of convention: end error messages with a period.

Also I believe in this particular message `(x, y, z, w)` information is irrelevant.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,74 @@
+            # Warn user for zero (eps?) norm and set to identity quaternion, which
+            # is (0,0,0,1) in (x,y,z,w) format
+            bool_indices = norms == 0
+            if np.any(bool_indices == True):
```
Just without `True`.

Also `bool_indices` -> `zero_norm(s?)` or something like that.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,74 @@
+            bool_indices = norms == 0
+            if np.any(bool_indices == True):
+                warnings.warn("Found zero norm quaternions in input, replacing with identity quaternions")
+                quat[bool_indices] = np.array([0, 0, 0, 1])
```
We don't want to modify `quat` and we want to do as little operations as possible. Figure out how to do that.

I guess we need to copy `quat` when normalizing and then do in-place division. This way we don't even need to introduce another variable name.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,74 @@
+        quat : array_like, shape (N, 4) or (4,)
+               Each row is a (possibly non-unit norm) quaternion in scalar-last
+               (x, y, z, w) format.
+
```
I believe there are typically no blank lines such this in scipy doctrings.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,74 @@
+        Parameters
+        ----------
+        quat : array_like, shape (N, 4) or (4,)
+               Each row is a (possibly non-unit norm) quaternion in scalar-last
```
It should be standard 4 spaces indentation.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,74 @@
+               Each row is a (possibly non-unit norm) quaternion in scalar-last
+               (x, y, z, w) format.
+
+        normalized : boolean, optional, default = False
```
This `default = ` while might be useful, but somehow isn't used in scipy docstrings. There are pros and cons to that, but sometimes you just can't briefly state what is default. 

So just write `Default is False.` or `False by default.` in the docstring (like the last sentence).

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,74 @@
+                     quaternions all have unit norm and are not normalized again.
+
+        """
+
```
Again I would advocate agains such blank lines. But if you strongly prefer it, you can add them.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -0,0 +1,30 @@
+    r = Rotation.from_quaternion(x)
+    expected_quat = x / np.array([[5], [13]])
+    assert_array_almost_equal(r._quat, expected_quat)
+
```
I believe PEP8 prescribes to use 2 blank lines between standalone functions and classes (and one between methods in classes). It is just sort of a convention.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,74 @@
+
+    @classmethod
+    def from_quaternion(cls, quat, normalized=False):
+        """Initialize Rotation from possibly unnormalized quaternions.
```
This is a style thing, but I guess you should simply state `Initialize (or Create?) Rotation from quaternions`. And below you can explain that the quaternions will be normalized depending on `normzlied` flag.

I guess we may leave style things for later.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,74 @@
+    Methods
+    -------
+    from_quaternion
+
```
My preference is to avoid such blank lines, but if you like it better --- you choice.

nmayorov reviewed (commented):
Some final remarks for this section.

adbugger committed:
Reduced number of operations when normalizing. Added tests.

adbugger commented:
Reduced number of operations when normalizing. Copied argument `quat` to `self._quat`, performed in-place division only on those quaternions which have non-zero norm, and replacement only on those which have zero norm.

I have also added some tests on non-standard input where I am catching errors with `pytest.raises(ValueError)`. I added a test case for zero-norm inputs; in that test case the `pytest.raises(UserWarning)` is present since warnings are being treated as errors by `scipy.spatial.transform.test()`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,75 @@
+        # Try to convert to numpy array
+        quat = np.asarray(quat, dtype=float)
+
+        if quat.ndim not in [1, 2]:
```
I thought about it a bit more and I believe the clearest way to check it would be
```
if quat.ndim not in [1, 2] or quat.shape[-1] != 4:
    raise ValueError("`quat` should be of shape (4,) or (N x 4).")
```
It has the advantage of being in one place.

Notice I used `` `quat` `` --- this is another (sort of) convention to mention input arguments in error messages. Use it if you like it, it makes sense to me.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,75 @@
+        if quat.shape[1] != 4:
+            raise ValueError("A quaternion should have 4 numbers.")
+
+        self._quat = quat
```
Well, it only assigns a reference. I guess you need to do `quat = quat.copy()` in `not normalized` branch and then `self._quat = quat`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,75 @@
+        quat : array_like, shape (N, 4) or (4,)
+            Each row is a (possibly non-unit norm) quaternion in scalar-last
+            (x, y, z, w) format.
+
```
This blank line is really unconventional for scipy codebase.

adbugger replied:
My bad. I removed about the blanks in the end and forgot about the ones in the middle.

nmayorov reviewed (commented):
I guess this should be final.

Also, add `setup.py` in `spatial.transform`, see `signal.windows`.

adbugger committed:
Added `transform/setup.py`, modified input shape check for clarity.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,71 @@
+            quat = quat[None, :]
+            self._single = True
+
+        self._quat = quat
```
Really nitpicky, but it seems a bit illogical. I would write this inside `if normalized`, and `if not normalized` will go as `else`.

adbugger committed:
Minor change to if-else logic

nmayorov commented:
I'm merging this.

A note about commits. It is preferable to solve one problem with one commit. For example, it's not very logical to mix `setup.py` creation with some algorithm logic fixes. And I bring up "commit tags" from numpy development guide again.

It's not a big problem. We will rebase all commits with proper messages later on. But better to think about it now.

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,71 @@
 
 import numpy as np
+import scipy.linalg
+import warnings
```
PEP8 two newlines before class def

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,71 @@
+        quat = np.asarray(quat, dtype=float)
+
+        if quat.ndim not in [1, 2] or quat.shape[-1] != 4:
+            raise ValueError("`quat` should be of shape (4,) or (N x 4).")
```
would be nice to add `, got shape %s' % (quat.shape,)`

nmayorov replied:
I support this. I prefer `.format(...)` though, @adbugger --- pick what you like more.

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,71 @@
+            zero_norms = norms == 0
+            if zero_norms.any():
+                warnings.warn("Found zero norm quaternions in input, replacing with identity quaternions.")
+                self._quat[zero_norms] = np.array([0, 0, 0, 1])
```
I'd rather just raise an error in this case.

nmayorov replied:
I thought about it too. I'm neutral (aka `+0`) to this and we have +1 to change from @larsoner.  

@adbugger --- what is your opinion?

adbugger replied:
Since rotations require unit norm quaternions, I think we can expect quaternions from even noisy measurements to not deviate so far from unit norm as to approach zero. I'd be fine with raising a `ValueError` (or some other appropriate error) in this case. @nmayorov does that sound fine?

nmayorov replied:
As I said, I think both options are reasonable. Perhaps I like throwing an error slightly more as well. So go ahead and change it to the error.

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1,12 +1,71 @@
     """
-    def __init__(self):
-        pass
+    def __init__(self, quat, normalized=False):
```
One other thing we could consider is supporting 3-element `(x,y,z)` quaternions if `normalized=True`, because `w` can be computed in this case.

nmayorov replied:
It is a bit tricky: you can't determine the sign of the last element, which is crucial. So we have to pass 4 numbers always, even though we know their sum of squares is 1.

larsoner replied:
We can say that we will always use the positive variant of the value. If people want something else they can provide it.

adbugger replied:
[__EDIT__: didn't see the above comment before typing]
Right, `w=1` and `w=-1` both represent the same overall rotation, but `w=-1` represents a rotation "the longer way around". While not important right now, this will have a major effect when we implement interpolation algorithms later on.

We could do a default of `w=1` or something, along with appropriate warnings, thoughts on that?

nmayorov replied:
To be honest, I think it is very unconventional to represent rotation quaternions as 3 numbers. I never saw them used that way or implemented that way. I think it will be confusing to allow such representation, I just don't see where it can occur.

larsoner replied:
I have used a program that had this convention, but if you haven't seen it at all then it's probably not widely used enough to do here. We can always add it later, too, if it turns out others also have seen it.

nmayorov replied:
Interesting. I'm pretty sure it's not widely used.

Let me also explain, why I think it is tricky and (sort of) bug prone. If we have 4 number `(x, y, z, w)` then in order to convert them to 3 numbers with a convention that `w > 0` we first need to check the sign of `w` and if it's negative --- negate `x, y, z`. If we forget to do that and just throw away `w`, then we will get a completely different rotation. And considering that all the algorithms with quaternions are conveniently expressed with `(x, y, z, w)` it seems indeed better to just keep `(x, y, z, w)` and don't use the sign tricks. In other words `(x, y, z, w)` representation is consistent and transparent, whereas `(x, y, z)` relies on an assumption, which is not really a part of quaternion mathematics.

And what Aditya said about interpolation --- this is quite interesting view as well. I didn't think about it and it is not always applicable, i.e. it is often said that `q` and `-q` are the same rotations. But looks like in fact `q` and `-q` might be viewed differently for interpolation problems, which is great.

So I suggest not to support such representation for now. Perhaps ask on the mailing list when everything will be more ready.

larsoner reviewed (commented):
Looks like a good start. Please run with a PEP8 checker, though.

adbugger commented:
@larsoner is there a preferred PEP8 checker for scipy? If not, how does `flake8` sound?

nmayorov commented:
@adbugger actually there is a packaged called `pep8` --- start with it, can use `flake8` as well.

larsoner commented:
Either is fine.

One other thing to do ASAP is to enable Travis-CI for your fork of SciPy. IIRC it includes these checks.
